Marlin/src/feature/bedlevel/bdl/bdl.cpp

```diff
@@ -54,139 +57,188 @@ BDS_Leveling bdl;
+          ui.set_status(tmp_1, true);
+        }
+      #endif
     }
     else
       stepper.apply_directions();
 
-    #if ENABLED(DEBUG_OUT_BD)
-      SERIAL_ECHOLNPGM("BD:", tmp & 0x3FF, ", Z:", cur_z, "|", current_position.z);
-      if (BD_I2C_SENSOR.BD_Check_OddEven(tmp) == 0) SERIAL_ECHOLNPGM("errorCRC");
-    #endif
+    DEBUG_ECHOLNPGM("BD:", tmp & 0x3FF, " Z:", cur_z, "|", current_position.z);
+    if (TERN0(DEBUG_OUT_BD, BD_I2C_SENSOR.BD_Check_OddEven(tmp) == 0)) DEBUG_ECHOLNPGM("CRC error");
 
-    if ((tmp & 0x3FF) > 1020) {
+    if ((tmp & 0x3FF) >= 1015) {
       BD_I2C_SENSOR.BD_i2c_stop();
       safe_delay(10);
     }
 
+    // Read version. Usually used as a connection check
+    if (config_state == -1) {
+      BD_I2C_SENSOR.BD_i2c_write(CMD_READ_VERSION);
+      safe_delay(100);
+      char tmp_1[21];
+      for (int i = 0; i < 19; i++) {
+        tmp_1[i] = BD_I2C_SENSOR.BD_i2c_read() & 0xFF;
+        safe_delay(50);
+      }
+      config_state = 0;
+      BD_I2C_SENSOR.BD_i2c_write(CMD_END_READ_CALIBRATE_DATA);
+      SERIAL_ECHOLNPGM("BD Sensor version:", tmp_1);
+      if (tmp_1[0] != 'V') SERIAL_ECHOLNPGM("Read Error. Check connection and delay.");
+      safe_delay(50);
+    }
     // read raw calibrate data
-    if (config_state == -5) {
+    else if (config_state == -5) {
       BD_I2C_SENSOR.BD_i2c_write(CMD_START_READ_CALIBRATE_DATA);
-      safe_delay(1000);
+      safe_delay(100);
 
       for (int i = 0; i < MAX_BD_HEIGHT * 10; i++) {
         tmp = BD_I2C_SENSOR.BD_i2c_read();
-        SERIAL_ECHOLNPGM("Calibrate data:", i, ",", tmp & 0x3FF, ", check:", BD_I2C_SENSOR.BD_Check_OddEven(tmp));
-        safe_delay(500);
+        SERIAL_ECHOLNPGM("Calibrate data (", i, ") = ", tmp & 0x3FF);
+        (void)check(tmp, i == 0);
+        safe_delay(50);
       }
       config_state = 0;
       BD_I2C_SENSOR.BD_i2c_write(CMD_END_READ_CALIBRATE_DATA);
-      safe_delay(500);
+      safe_delay(50);
     }
     else if (config_state <= -6) {   // Start Calibrate
-      safe_delay(100);
+      safe_delay(10);
       if (config_state == -6) {
-        //BD_I2C_SENSOR.BD_i2c_write(1019); // begin calibrate
-        //delay(1000);
-        gcode.stepper_inactive_time = SEC_TO_MS(60 * 5);
-        gcode.process_subcommands_now(F("M17 Z"));
-        gcode.process_subcommands_now(F("G1 Z0.0"));
-        z_pose = 0;
+        REMEMBER(gsit, gcode.stepper_inactive_time, SEC_TO_MS(60 * 5));
+        SERIAL_ECHOLNPGM("c_z0:", planner.get_axis_position_mm(Z_AXIS), "-", pos_zero_offset);
+
+        // Move the z axis instead of enabling the Z axis with M17
+        // TODO: Use do_blocking_move_to_z for synchronized move.
+        current_position.z = 0;
+        sync_plan_position();
+        gcode.process_subcommands_now(F("G1Z0.05"));
         safe_delay(1000);
+        gcode.process_subcommands_now(F("G1Z0.00"));
+        safe_delay(1000);
+        current_position.z = 0;
+        sync_plan_position();
+        //safe_delay(1000);
```

Comment:
What is the exact purpose of either the `M17 Z` or the move and long delay here? I bet there is a better way to accomplish it with some direct internal calls rather than running G-code.

Reply:
it is the preparation for calibration after user have put the nozzle close to the bed, need to enable the z motors but I find that the z motor will move a little from unlock status to lock status that sometime will let the nozzle too close to the bed and will cause the first calibration data not good.

Reply:
Are the two one-second delays necessary? It's not the longest delay in the world, but a minimal delay is preferred.

Meanwhile, it appears that this could be called at any time after `config_state` is set to -6 or -7, and `current_position.z = 0; sync_plan_position();` causes Z to be set to 0 wherever it happens to be, so there's no obvious guarantee that it's near the bed. Since the Z position is corrupted by this activity, there should be some assurance that the calibration is followed by `G28 Z` at minimum. Of course, it is best if the Z position is not forcibly altered, if possible.